zmij.cc

```diff
@@ -736,13 +782,13 @@ inline void write2digits(char* buffer, uint32_t value) noexcept {
   memcpy(buffer, digits2(value), 2);
 }
 
-// Writes 4 digits and removes trailing zeros.
-auto write4digits_trim_zeros(char* buffer, uint32_t value) noexcept -> char* {
-  auto [aa, bb] = divmod100(value);
-  write2digits(buffer + 0, aa);
-  write2digits(buffer + 2, bb);
-  return buffer + 4 - num_trailing_zeros[bb] -
-         (bb == 0) * num_trailing_zeros[aa];
+auto write8digits(uint32_t aa, uint32_t bb, uint32_t cc, uint32_t dd) -> uint64_t {
+  union { char b[8]; uint64_t u64; } bits;
+  write2digits(&bits.b[0], aa);
+  write2digits(&bits.b[2], bb);
+  write2digits(&bits.b[4], cc);
+  write2digits(&bits.b[6], dd);
+  return bits.u64;
```

Comment:
I'm pretty sure this is a UB in C++.

Reply:
Me too! I was looking for a reasonably clean form that guarantees that the value is materialized in a register, and couldn't find an option that doesn't look like
   memset((char *)some_uint64 + 2, digits(bb), 2)
which I'm fairly sure is undefined as well, no matter how I precisely address the bit range in the uint64_t. Anyway, I think I came up with something that works and is not actually undefined, I just need to confirm that the assembly looks reasonable.

Reply:
I changed this to use memcpy to uint16_t variables which are then shifted and or'ed to yield the final uint64_t. The assembly ended up becoming a bit shorter as well.